### DISCUSSION TITLE
"Unable to resolve type" for simple UserControl

### DISCUSSION THREAD

**Discussion by sedwick2048** [AvaloniaUI/Avalonia#16049] [Q&A] [answered]
I'm working on my first big Avalonia project, attempting to convert a small WPF app.  Had some success, then brought in some more user controls and started getting dozens of errors, most of them understandable given how Avalonia does things a little differently and I had yet to switch to the new way in many places.  

I decided to see if I could create a whole new user control using Visual Studio 2022's "add new item" feature, instead of porting one in from the WPF app, and I made some changes to initialize it more like the other successful UCs.  But like the failing ones, it gives me "Unable to resolve type MyProject.Views.TestUserControl", even though there's hardly anything in it:

```
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
             x:Class="MyProject.Views.TestUserControl">
  <Grid>
    <TextBlock Text="Welcome to Avalonia!" />
  </Grid>
</UserControl>
```

Code behind:
```
using Avalonia.Controls;
using Avalonia.Markup.Xaml;

namespace MyProject.Views {
  public partial class TestUserControl : UserControl {
    public TestUserControl() {
      AvaloniaXamlLoader.Load(this);
    }
  }
}
```

Additionally, the design preview of every view including MainWindow has an exclamation point in a red circle and "Build Required...Could not find 'C:\Users\Me\source\repos\MyProject\obj\Debug\net8.0\ref\MyProject.dll'.  Please build your project to enable previewing and Intellisense."  And of course I can't build yet.

Any ideas how to solve this?  Thanks...

**Comment by rabbitism** [comment]
What new item did you add? Avalonia UC or WPF UC?

**Reply by sedwick2048** [reply]
Pretty sure "User Control (Avalonia)" was my only option.  At least, it was after I updated my VS ten minutes ago.

**Comment by IanRawley** [comment]
That AvaloniaXamlLoader.Load(this); looks a bit suspect to me.  My UserControl derived controls have that replaced with InitializeComponent();

What version of the Avalonia packages are you using?

Edit:  Actually, also make sure your Visual Studio extension is up to date.

**Reply by sedwick2048** [reply]
I switched in an `InitializeComponent` call and it didn't change anything.  My VS is set to update extensions automatically.  Oddly, despite Avalonia's influence on many VS interactions (like selecting a user control for a new item), VS Installer makes no mention of Avalonia in "Workloads" or "Individual Components"...

**Reply by reelthymeos** [reply]
My Avalonia package is the most recent (11.0.11) and I am getting a bunch of errors like this too. All I did was create a desktop project and try to create a few user controls, but there were many such errors in the output and the designer would not display anything. As someone else has suggested, a way around this is to simply create a new project and copy your code over. That's what I did and for whatever reason, it now works. 

Try the following:

* Close and then relaunch VS
* Create a new project
* Copy your original code over

****UPDATE: I now created a new user control and I'm having the same issue. As soon as the control is created, the designer tells me that the markup is invalid and I get the error "unable to resolve type myapp.Views.MyClass". _However, as soon as I build the project, the issue resolves._  ****

**Comment by sedwick2048** [comment, marked as answer]
Well, I upgraded Visual Studio to v17.10.2 and this error pretty much disappeared.  I worked through all my other errors (listed one at a time after each build, ugh) and now I have something that runs.  Only two of my user controls have "Invalid Markup", one of which says it "Cannot find an event named PointerEnter on type StackPanel" for my EventTriggerBehavior, where my WPF app could have an EventTrigger with EventName="MouseEnter".  The other control has the same error, probably because it's the parent control.  New discussion, I guess...

**Reply by sedwick2048** [reply]
UPDATE: the correct event was "PointerEntered".  After a clean and rebuild, I have designer previews now!

**Comment by NickBarton** [comment]
I get this problem each time I create a new UserControl.

My usual problem is not getting the namespace names correct (I'm layering my project according to use cases and not Views / Models).

Then a clean build of the whole solution (not just the UI project) solves it. Such a pain as the solution is large.

**Reply by markterluun** [reply]
Rebuilding the entire solution seems to solve the problem for me as well, a huge pain indeed.
